library/data-model/src/data_storage/internals.ts

```diff
@@ -803,6 +804,126 @@ export async function getRecord({
   });
 }
 
+/**
+ * Creates a canonical hash from the record audit data for efficient sync checking.
+ *
+ * @param audit - Array of {id, rev} objects from getRecordAudit
+ * @returns SHA-256 hash of the canonical representation
+ */
+async function createAuditHash(
+  audit: Array<{id: string; rev: string}>
+): Promise<string> {
+  // The audit is already sorted by ID in getRecordAudit, but ensure consistency
+  const sortedAudit = audit
+    .slice() // Don't mutate original
+    .sort((a, b) => a.id.localeCompare(b.id));
+
+  // Create canonical JSON representation
+  const canonicalJson = JSON.stringify(sortedAudit, null, 2);
+
+  // console.log('canonical json', canonicalJson);
+  // Return SHA-256 hash
+  return await createHash(canonicalJson);
+}
+
+/**
+ * Generate an audit for a set of records.
+ * Based on a list of all document ids and revision ids for documents
+ * making up each record.  All frevs, avps but not attachment documents since
+ * they may not be present on the client.
+ * We then create a hash of the audit data to use for sync checking.
+ *
+ * We can then use this in the client to verify that we have a complete and
+ * up to date copy of the record on the server.
+ *
+ */
+export type RecordAuditMap = {
+  [recordId: string]: string;
+};
+
+export async function getRecordListAudit({
+  recordIds,
+  dataDb,
+}: {
+  recordIds: RecordID[];
+  dataDb: DataDbType;
+}): Promise<RecordAuditMap> {
+  // Early return for empty input
+  if (!recordIds || recordIds.length === 0) {
+    return {};
+  }
+
+  // Batch fetch all records at once
+  const recordsResult = await dataDb.find({
+    selector: {
+      _id: {
+        $in: recordIds,
+      },
+    },
+    fields: ['_id', '_rev'],
+    limit: recordIds.length + 10,
+  });
+
+  // Batch fetch all related documents (revisions, AVPs) in one query
+  const relatedDocsResult = await dataDb.find({
+    selector: {
+      record_id: {
+        $in: recordIds,
+      },
+      attach_format_version: {
+        $exists: false,
+      },
+    },
+    fields: ['_id', '_rev', 'record_id'],
```

Comment:
can you highlight this with a comment - important that we only look at id, rev, record_id as the fields being hashed